### PR TITLE
Make modules with tests runnable

### LINKED ISSUE #154: Run all tests in the module
We have a feature to invoke the right `cargo test` command for the test at caret:

image

(In VS Code, it is triggered by the **Rust Run** action)

Currently, it works only for `#[test]` functions, but it would be useful to run test modules as well.

Here's how the feature works currently:

`runnables` function discovers `#[test]` functions: 
https://github.com/rust-analyzer/rust-analyzer/blob/0043d7e9c7f27af95cbf299d5b92ce28ba9945ab/crates/ra_editor/src/lib.rs#L119-L140 

The function is exposed on `Analysis` struct (the API between language server and "compiler") here: 
https://github.com/rust-analyzer/rust-analyzer/blob/0043d7e9c7f27af95cbf299d5b92ce28ba9945ab/crates/ra_analysis/src/lib.rs#L245-L248

in `main_loop/handlers` we add information about current Cargo target/package to it and handle, as JSON, to the editor:

https://github.com/rust-analyzer/rust-analyzer/blob/0043d7e9c7f27af95cbf299d5b92ce28ba9945ab/crates/ra_lsp_server/src/main_loop/handlers.rs#L255-L268

To support testing *modules*, we need to add fully-qualified module name to the command line. The steps to do so would be

* move the `runnable` function from ra_editor to ra_analysis: ra_editor knows nothing about module
* in the `runnable` function, detect not only functinons, but mods with test functions as well
* make a function to compute a `foo::bar::baz` path to a module. See the implementation  of  [crate_root](https://github.com/rust-analyzer/rust-analyzer/blob/0043d7e9c7f27af95cbf299d5b92ce28ba9945ab/crates/ra_analysis/src/lib.rs#L242-L244) function, which contains "traverse parent modules" logic. 
* add module path to a `Runnable` struct
* adjust the `handle_runnables` function to construct a command line, using the full path

### PR DESCRIPTION
**Pull request by farodin91**
Fixes #154

I having problems to traverse the path to module. The main problem is that module_tree only supports `FileId` and not `Module` in files. Any idea?

I need to clean up the code a bit later.

### CODE REVIEW

**Comment by matklad** [comment]
> The main problem is that module_tree only supports FileId and not Module in files. 

Yeah, it indeed does not handle modules inside the file for the time being, so it's ok to just write custom code for modules in files. In the future, we need to modify module tree to work uniformly with all modules.

**Comment by farodin91** [comment]
I would like to making the command line a bit better by adding the package. Any idea? To get crate name of a file.

**Comment by matklad** [comment]
>I would like to making the command line a bit better by adding the package.

I think this line should add the neccessary `--package foo --test bar` arguments:
https://github.com/rust-analyzer/rust-analyzer/pull/165/files#diff-57a6a05c79db4a0461f95762e92206d3R314

**Comment by farodin91** [comment]
If i'm correct line https://github.com/rust-analyzer/rust-analyzer/pull/165/files#diff-57a6a05c79db4a0461f95762e92206d3R310 already add the package.

**Comment by matklad** [comment]
@farodin91 yep, I've linked the wrong line

**Comment by matklad** [comment]
@farodin91 https://github.com/rust-analyzer/rust-analyzer/pull/193 adds inline modules to the module tree

**Comment by aochagavia** [comment]
Besides rebasing, what needs to happen so this PR can be merged?

**Comment by farodin91** [comment]
The rebase is a bit complexer.

**Comment by farodin91** [comment]
@matklad The module parent function seems to behave wrong. **test_resolve_parent_module_for_inline_to_lib** it returns the current mod.

File: crates/ra_analysis/src/imp.rs [hunk 1]
```diff
@@ -220,6 +220,34 @@ impl AnalysisImpl {
         let source_root = self.db.file_source_root(file_id);
         self.db.module_tree(source_root)
     }
+    pub(crate) fn module_path(&self, position: FilePosition) -> Cancelable<String> {
+        let mut current_position = position;
+        let mut path = "".to_string();
+
+        loop {
+            println!("current_position {:?}", current_position);
+            if let Some((file_id, file_symbol)) = self.parent_module(current_position)?.first() {
```
Comment (matklad) [inline crates/ra_analysis/src/imp.rs]: Hm, calling `Analysis::parent_module` is probably not the right thing to do here: this function is an API function used by LSP to implement "go to super module" feature. 

We should add a `path_to_root` method to `ModuleId` instead, here:

https://github.com/rust-analyzer/rust-analyzer/blob/923483e321acace3bbf38688bd70d4d38f49b35e/crates/ra_analysis/src/descriptors/module/mod.rs#L82

then, we should use this method directly here, but constructing ModuleSource, converting it to ModuleId, etc, like we do in the implementation of Analysis::parent_module

Comment (matklad) [reply crates/ra_analysis/src/imp.rs]: > this function is an API function used by LSP to implement "go to super module" feature.

This explains why, for inline modules, it returns the location of module's own identifier: this mirrors the way it woks for non-inline modules.

Comment (farodin91) [reply crates/ra_analysis/src/imp.rs]: I try to get a the module name by the ModuleId and it seems a bit problematic.

Comment (matklad) [reply crates/ra_analysis/src/imp.rs]: Take a look at #234: it introduces a more OO-flavored API for modules, which hopefully should be easier to use.

In particular, a `ModuleDescriptor` now has  a `::name` method.

**Comment by matklad** [comment]
Oups, @farodin91 sorry for the long delay here: GitHub does not send notifications for force-pushes, so I've missed the latest developments. Will take a look soom!

File: crates/ra_analysis/src/imp.rs [hunk 2]
```diff
@@ -184,6 +184,28 @@ impl AnalysisImpl {
         };
         Ok(query.search(&buf))
     }
+
+    pub(crate) fn module_path(&self, position: FilePosition) -> Cancelable<String> {
```
Comment (DJMcNab) [inline crates/ra_analysis/src/imp.rs]: I feel like this should return `Cancelable<Option<String>>` as opposed to returning `Ok("")` in the failure cases.

File: crates/ra_analysis/src/imp.rs [hunk 3]
```diff
@@ -184,6 +184,28 @@ impl AnalysisImpl {
...
+            None => return Ok("".to_string()),
+            Some(it) => it.to_string(),
+        };
+
+        let modules = descr.path_to_root();
+
+        let path = modules
+            .into_iter()
+            .filter_map(|s| s.name())
+            .skip(1)
```
Comment (DJMcNab) [inline crates/ra_analysis/src/imp.rs]: I'm sure it's there for a reason, but a comment explaining what `skip(1)` does in this case would be useful.

File: crates/ra_analysis/src/imp.rs [hunk 4]
```diff
@@ -184,6 +184,28 @@ impl AnalysisImpl {
...
+        };
+        let name = match descr.name() {
+            None => return Ok("".to_string()),
+            Some(it) => it.to_string(),
+        };
+
+        let modules = descr.path_to_root();
+
+        let path = modules
+            .into_iter()
```
Comment (matklad) [inline crates/ra_analysis/src/imp.rs]: Changing this `.into_iter()` into `.iter()` should fix the borrowing error.

File: crates/ra_analysis/src/imp.rs [hunk 5]
```diff
@@ -181,6 +181,28 @@ impl AnalysisImpl {
...
+            None => return Ok(None),
+            Some(it) => it.to_string(),
+        };
+
+        let modules = descr.path_to_root();
+
+        let path = modules
+            .iter()
+            .filter_map(|s| s.name())
+            .skip(1) // name is already part of the string.
```
Comment (DJMcNab) [inline crates/ra_analysis/src/imp.rs]: Thanks!

**Comment by DJMcNab** [comment]
bors delegate+

This looks good - all that's missing is you have an extra import.

**Pull request merged**